core/include/cubos/core/memory/unordered_bimap.hpp

```diff
@@ -0,0 +1,164 @@
+    class UnorderedBimap final
+    {
+    public:
+        using Iterator = typename std::unordered_map<L, R, LHash>::const_iterator;
+
+        /// @brief Adds a new entry to the map.
+        /// @note If any of the values already exists, the old entries with them are removed.
+        /// @param left Left value.
+        /// @param right Right value.
+        void add(L left, R right)
+        {
+            auto leftIt = mLeftToRight.find(left);
+            if (leftIt != mLeftToRight.end())
+            {
+                mRightToLeft.erase(leftIt->second);
+            }
+
+            auto rightIt = mRightToLeft.find(right);
+            if (rightIt != mRightToLeft.end())
+            {
+                mLeftToRight.erase(rightIt->second);
+            }
+
+            mLeftToRight.insert_or_assign(left, right);
+            mRightToLeft.insert_or_assign(right, left);
+        }
+
+        /// @brief Removes the entry associated to the given left value.
+        /// @param left Left value.
+        /// @return Whether the entry was removed.
+        bool removeLeft(const L& left)
+        {
+            auto it = mLeftToRight.find(left);
+            if (it == mLeftToRight.end())
+            {
+                return false;
+            }
+
+            mRightToLeft.erase(it->second);
+            mLeftToRight.erase(it);
+            return true;
+        }
+
+        /// @brief Removes the entry associated to the given right value.
+        /// @param right Right value.
+        /// @return Whether the entry was removed.
+        bool removeRight(const R& right)
+        {
+            auto it = mRightToLeft.find(right);
+            if (it == mRightToLeft.end())
+            {
+                return false;
+            }
+
+            mLeftToRight.erase(it->second);
+            mRightToLeft.erase(it);
+            return true;
+        }
+
+        /// @brief Checks if the map has the given entry.
+        /// @param left Left value.
+        /// @param right Right value.
+        /// @return Whether the map has the entry.
+        bool has(const L& left, const R& right) const
+        {
+            auto it = mLeftToRight.find(left);
+            if (it == mLeftToRight.end())
+            {
+                return false;
+            }
+            return it->second == right;
+        }
+
+        /// @brief Checks if the map contains the given left value.
+        /// @param left Left value.
+        /// @return Whether the map contains the value.
+        bool hasLeft(const L& left) const
+        {
+            return mLeftToRight.find(left) != mLeftToRight.end();
+        }
+
+        /// @brief Checks if the map contains the given right value.
+        /// @param right Right value.
+        /// @return Whether the map contains the value.
+        bool hasRight(const R& right) const
+        {
+            return mRightToLeft.find(right) != mRightToLeft.end();
+        }
```

Comment:
Why use `.find` and not `.contains`

Reply:
Assumed not all compilers would support it for some reason. will switch